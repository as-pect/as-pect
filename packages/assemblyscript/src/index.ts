export type ASC = typeof import("assemblyscript/dist/asc.js") ;
export type AS = typeof import("assemblyscript/dist/assemblyscript.js");
export type TRANSFORM = typeof import("assemblyscript/dist/transform.js");

// @ts-ignore: Adding an aspect global
if (!globalThis.__aspect) globalThis.__aspect = {};

// We need to garuntee we use only a single version of the compiler

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.asc) globalThis.__aspect.asc = await import("assemblyscript/dist/asc.js");
// @ts-ignore: Adding an aspect global for asc
export const asc: ASC = globalThis.__aspect.asc;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.assemblyscript) globalThis.__aspect.assemblyscript = await import("assemblyscript/dist/assemblyscript.js");
// @ts-ignore: Adding an aspect global for asc
export const assemblyscript: AS = globalThis.__aspect.assemblyscript;

// @ts-ignore: Adding an aspect global for asc
if (!globalThis.__aspect.transform) globalThis.__aspect.transform = await import("assemblyscript/dist/transform.js");
// @ts-ignore: Adding an aspect global for asc
export const transform: TRANSFORM = globalThis.__aspect.transform;

