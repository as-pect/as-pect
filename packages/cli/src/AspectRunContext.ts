import { IAspectConfiguration } from "./IAspectConfiguration";
import { IProcessConfiguration } from "./IProcessConfiguration";
import { IWritable } from "@as-pect/core/src/util/IWritable";

export interface IASCProps {
  stdout: IWritable;
  stderr: IWritable;
  listFiles(dirname: string, baseDir: string): string[];
  readFile(filename: string, baseDir: string): Uint8Array | null;
  writeFile(name: string, contents: Uint8Array, baseDir?: string): void;
}

export interface IASC {
  ready(): Promise<void>;
  main(
    args: string[],
    config: IASCProps,
    callback: (error: Error) => number,
  ): void;
}


export interface IOptionalRunContextProps {
  /** The assemblyscript help text generation function. */
  help: typeof import("assemblyscript/cli/util/options").help;
  /** The assemblyscript cli parse function. */
  parse: typeof import("assemblyscript/cli/util/options").parse;
  /** The result of calling the assemblyscript cli parse function. */
  parseResult: import("assemblyscript/cli/util/options").Result;
  /** The retreived as-pect config file. */
  config: Partial<IAspectConfiguration>;
  /** The assemblyscript compiler cli interface itself. */
  asc: IASC;
}

export interface IRequiredRunContextProps {
  /** Raw argv from the node process. */
  argv: string[];
  /** A set of file promises that must be resolved before the application exits. */
  files: Promise<void>[];
  /** A stripped down version of a process that contains an exit function, stdout and stderr. */
  process: IProcessConfiguration;
  /** The exit code used to exit the program. It starts as 0, and can only be set to 1 otherwise. */
  exitCode: number;
}

export type AspectRunContext = Partial<IOptionalRunContextProps> &
  IRequiredRunContextProps;
