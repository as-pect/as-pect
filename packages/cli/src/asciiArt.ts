import chalk from "chalk-template";

/**
 * @ignore
 *
 * This method prints the ascii art.
 * @param {string} version - The cli version
 */
export function printAsciiArt(): void {
  console.log(chalk`{bold.bgWhite.black ${""}       ___   _____                       __    
      /   | / ___/      ____  ___  _____/ /_   
     / /| | \\__ \\______/ __ \\/ _ \\/ ___/ __/   
    / ___ |___/ /_____/ /_/ /  __/ /__/ /_     
   /_/  |_/____/     / .___/\\___/\\___/\\__/     
                    /_/                        }
`);
}
