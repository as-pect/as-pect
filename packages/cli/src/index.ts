import { parse } from "configinator";
import { cliConfig, resolveOptionByName } from "./util/configuration";
import chalk from "chalk";
const pkg = require("../package.json");

export function asp(args: string[], readFileSync: (file: string, basename: string) => string | null): void {
  const configState = parse(args, cliConfig, {
    cwd: process.cwd(),
    readFileSync,
  });

  if (configState.diagnostics.length > 0) {
    for (const diag of configState.diagnostics) {
      console.log(chalk`{red [CLI Error]}: ${diag}`);
    }
    process.exit(1);
  }

  if (resolveOptionByName(configState, "version")) {
    console.log(chalk`{yellow [Version]}: ${pkg.version}`);
    process.exit(0);
  }

  if (!resolveOptionByName(configState, "nologo")) {
    const printAsciiArt = require("./util/asciiArt").printAsciiArt;
    printAsciiArt(pkg.version);
  }

  if (resolveOptionByName(configState, "types")) {
    const types = require("./types").types;
    types();
    process.exit(0);
  }

  if (resolveOptionByName(configState, "portable")) {
    const portable = require("./portable").portable;
    portable();
    process.exit(0);
  }

  process.exit(0);
}

if (typeof require != "undefined" && require.main == module) {
  const fs = require("fs");
  const path = require("path");

  asp(process.argv.slice(2), (file: string, basename: string): string | null => {
    const thePath = path.normalize(path.join(basename, file));
    try {
      return fs.readFileSync(thePath, "utf8");
    } catch (ex) {
      return null;
    }
  });
}