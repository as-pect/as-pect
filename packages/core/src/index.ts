export * from "./reporter/CombinationReporter";
export * from "./reporter/EmptyReporter";
export * from "./reporter/IReporter";
export * from "./reporter/SummaryReporter";
export * from "./reporter/VerboseReporter";
export * from "./test/IWarning";
export * from "./test/TestContext";
export * from "./test/TestNode";
export * from "./util/IAspectExports";
export * from "./util/IPerformanceConfiguration";
export * from "./util/ReflectedValue";
