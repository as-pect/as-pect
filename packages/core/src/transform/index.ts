// import { Transform } from "assemblyscript/cli/transform";
import {
  ClassDeclaration,
  NamespaceDeclaration,
  NodeKind,
  Parser,
  Statement,
  Transform,
  Source,
} from "./assemblyscript";
import { createStrictEqualsMember } from "./createStrictEqualsMember";
import { createAddReflectedValueKeyValuePairsMember } from "./createAddReflectedValueKeyValuePairsMember";
export = class AspectTransform extends Transform {
  /**
   * This method results in a pure AST transform that inserts a strictEquals member
   * into each ClassDeclaration.
   *
   * @param {Parser} parser - The AssemblyScript parser.
   */
  afterParse(parser: Parser): void {
    // For backwards compatibility
    let sources: Source[] = (parser as any).program
      ? (parser as any).program.sources
      : parser.sources;
    // for each program source
    for (const source of sources) {
      traverseStatements(source.statements);
    }
  }
};

function traverseStatements(statements: Statement[]): void {
  // for each statement in the source
  for (const statement of statements) {
    // find each class declaration
    if (statement.kind === NodeKind.CLASSDECLARATION) {
      // cast and create a strictEquals function
      const classDeclaration = <ClassDeclaration>statement;
      classDeclaration.members.push(createStrictEqualsMember(classDeclaration));
      classDeclaration.members.push(
        createAddReflectedValueKeyValuePairsMember(classDeclaration),
      );
    } else if (statement.kind === NodeKind.NAMESPACEDECLARATION) {
      const namespaceDeclaration = <NamespaceDeclaration>statement;
      traverseStatements(namespaceDeclaration.members);
    }
  }
}
