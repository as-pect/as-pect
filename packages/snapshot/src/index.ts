export * from "./parser";
export * from "./test/Snapshot";
export * from "./test/SnapshotDiff";
export * from "./test/unparse";
