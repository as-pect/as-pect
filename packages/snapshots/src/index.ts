export * from "./Snapshot";
export * from "./SnapshotDiff";
export * from "./SnapshotDiffResult";
