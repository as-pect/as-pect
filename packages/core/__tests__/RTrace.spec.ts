import { TestContext, TestNodeType, EmptyReporter } from "../src";
import { promises as fs } from "fs";
import { instantiate } from "assemblyscript/lib/loader/umd";

describe("RTrace output", () => {
  test("Overall Statistics", async () => {
    const binary = await fs.readFile("./assembly/jest-rtrace.wasm");
    const ctx = new TestContext({
      binary,
      reporter: new EmptyReporter(),
      fileName: "./assembly/jest-rtrace.ts",
    });
    ctx.run(await instantiate(binary, ctx.createImports({})));
    expect(ctx.freeCount).toMatchSnapshot("freeCount");
    expect(ctx.allocationCount).toMatchSnapshot("allocationCount");
    expect(ctx.errors).toMatchSnapshot("errors"); // matches empty array
    ctx.rootNode.visit((group) => {
      if (group.type === TestNodeType.Group) {
        expect(group.rtraceStart).toMatchSnapshot(`${group.name} rtraceStart`);
        expect(group.rtraceEnd).toMatchSnapshot(`${group.name} rtraceEnd`);
        expect(group.rtraceDelta).toMatchSnapshot(`${group.name} rtraceDelta`);
        expect(group.errors).toMatchSnapshot(`${group.name}errors`); // matches empty array
        for (const groupTest of group.groupTests) {
          expect(groupTest.rtraceStart).toMatchSnapshot(
            `${groupTest.namespace} rtraceStart`,
          );
          expect(groupTest.rtraceEnd).toMatchSnapshot(
            `${groupTest.namespace} rtraceEnd`,
          );
          expect(groupTest.rtraceDelta).toMatchSnapshot(
            `${groupTest.namespace} rtraceDelta`,
          );
          expect(groupTest.errors).toMatchSnapshot(
            `${groupTest.namespace} errors`,
          ); // matches empty array
        }
      }
    });
  });
});
