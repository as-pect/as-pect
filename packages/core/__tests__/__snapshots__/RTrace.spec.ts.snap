// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeBlocks()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeBlocks()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeBlocks()[0] rtraceEnd 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeBlocks()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeNodeBlocks()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeNodeBlocks()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeNodeBlocks()[0] rtraceEnd 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.activeNodeBlocks()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.allocations()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.allocations()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.allocations()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.allocations()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.count()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.count()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.count()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.count()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.decrements()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.decrements()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.decrements()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.decrements()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.frees()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.frees()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.frees()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.frees()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.increments()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.increments()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.increments()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.increments()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeAllocations()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeAllocations()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeAllocations()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeAllocations()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeDecrements()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeDecrements()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeDecrements()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeDecrements()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeFrees()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeFrees()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeFrees()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeFrees()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeIncrements()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeIncrements()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeIncrements()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeIncrements()[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeReallocations()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeReallocations()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeReallocations()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.nodeReallocations()[0] rtraceStart 1`] = `31`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.reallocations()[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.reallocations()[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.reallocations()[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.reallocations()[0] rtraceStart 1`] = `31`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOf(pointer: usize)[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOf(pointer: usize)[0] rtraceDelta 1`] = `4`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOf(pointer: usize)[0] rtraceEnd 1`] = `27`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOf(pointer: usize)[0] rtraceStart 1`] = `23`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOfReference<T>(reference: T)[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOfReference<T>(reference: T)[0] rtraceDelta 1`] = `4`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOfReference<T>(reference: T)[0] rtraceEnd 1`] = `31`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.refCountOfReference<T>(reference: T)[0] rtraceStart 1`] = `27`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOf(pointer: usize)[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOf(pointer: usize)[0] rtraceDelta 1`] = `4`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOf(pointer: usize)[0] rtraceEnd 1`] = `11`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOf(pointer: usize)[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOfReference<T>(reference: T)[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOfReference<T>(reference: T)[0] rtraceDelta 1`] = `4`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOfReference<T>(reference: T)[0] rtraceEnd 1`] = `15`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.sizeOfReference<T>(reference: T)[0] rtraceStart 1`] = `11`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOf(pointer: usize)[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOf(pointer: usize)[0] rtraceDelta 1`] = `4`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOf(pointer: usize)[0] rtraceEnd 1`] = `19`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOf(pointer: usize)[0] rtraceStart 1`] = `15`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOfReference<T>(reference: T)[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOfReference<T>(reference: T)[0] rtraceDelta 1`] = `4`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOfReference<T>(reference: T)[0] rtraceEnd 1`] = `23`;

exports[`RTrace output Overall Statistics: !~RTrace api[0]!~RTrace.typeIdOfReference<T>(reference: T)[0] rtraceStart 1`] = `19`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate a bunch of objects[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate a bunch of objects[0] rtraceDelta 1`] = `3`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate a bunch of objects[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate a bunch of objects[0] rtraceStart 1`] = `3`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate an object[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate an object[0] rtraceDelta 1`] = `1`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate an object[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should allocate an object[0] rtraceStart 1`] = `3`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should count references[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should count references[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should count references[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should count references[0] rtraceStart 1`] = `6`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate all the objects[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate all the objects[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate all the objects[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate all the objects[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate an object[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate an object[0] rtraceDelta 1`] = `-1`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate an object[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should deallocate an object[0] rtraceStart 1`] = `4`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should start a label[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should start a label[0] rtraceDelta 1`] = `1`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should start a label[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~should start a label[0] rtraceStart 1`] = `6`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~trigger a reallocation[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~trigger a reallocation[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~trigger a reallocation[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~RTrace tests[0]!~trigger a reallocation[0] rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: !~a group reference counting error[0]!~empty[0] errors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: !~a group reference counting error[0]!~empty[0] rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: !~a group reference counting error[0]!~empty[0] rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: !~a group reference counting error[0]!~empty[0] rtraceStart 1`] = `0`;

exports[`RTrace output Overall Statistics: RTrace api rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: RTrace api rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: RTrace api rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: RTrace apierrors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: RTrace tests rtraceDelta 1`] = `0`;

exports[`RTrace output Overall Statistics: RTrace tests rtraceEnd 1`] = `0`;

exports[`RTrace output Overall Statistics: RTrace tests rtraceStart 1`] = `3`;

exports[`RTrace output Overall Statistics: RTrace testserrors 1`] = `Array []`;

exports[`RTrace output Overall Statistics: a group reference counting error rtraceDelta 1`] = `-1`;

exports[`RTrace output Overall Statistics: a group reference counting error rtraceEnd 1`] = `7`;

exports[`RTrace output Overall Statistics: a group reference counting error rtraceStart 1`] = `7`;

exports[`RTrace output Overall Statistics: a group reference counting errorerrors 1`] = `
Array [
  Object {
    "message": "An orphaned dellocation has occurred at block: 6284",
    "stackTrace": "    at ~lib/rt/tlsf/freeBlock (wasm-function[47]:0x1b88)
    at ~lib/rt/tlsf/__free (wasm-function[78]:0x2467)
    at start:assembly/jest-rtrace~anonymous|1~anonymous|0 (wasm-function[79]:0x246e)
    at ../assembly/assembly/internal/call/__call (wasm-function[153]:0x3314)",
    "type": "Orphaned Deallocation Error",
  },
]
`;

exports[`RTrace output Overall Statistics: allocationCount 1`] = `51`;

exports[`RTrace output Overall Statistics: errors 1`] = `
Array [
  Object {
    "message": "An orphaned dellocation has occurred at block: 6284",
    "stackTrace": "    at ~lib/rt/tlsf/freeBlock (wasm-function[47]:0x1b88)
    at ~lib/rt/tlsf/__free (wasm-function[78]:0x2467)
    at start:assembly/jest-rtrace~anonymous|1~anonymous|0 (wasm-function[79]:0x246e)
    at ../assembly/assembly/internal/call/__call (wasm-function[153]:0x3314)",
    "type": "Orphaned Deallocation Error",
  },
]
`;

exports[`RTrace output Overall Statistics: freeCount 1`] = `21`;
